File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        self.count = 0
        self.result = None

        def inorder(curr):
            if not curr or self.result is not None:
                return
            inorder(curr.left)
            self.count += 1
            if self.count == k:
                self.result = curr.val
                return
            inorder(curr.right)
        
        inorder(root)
        return self.result

File: Data Structures & Algorithms/two-sum-bsts/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public boolean twoSumBSTs(TreeNode root1, TreeNode root2, int target) {
        Set<Integer> seen = new HashSet<>();
        collect(root1, seen);
        return search(root2, target, seen);
    }

    private void collect(TreeNode curr, Set<Integer> seen) {
        if (curr == null) {
            return;
        }
        seen.add(curr.val);
        collect(curr.left, seen);
        collect(curr.right, seen);
    }

    private boolean search(TreeNode curr, Integer target, Set<Integer> seen) {
        if (curr == null) {
            return false;
        }
        int compliment = target - curr.val;
        if (seen.contains(compliment)) {
            return true;
        }
        return search(curr.left, target, seen) || search(curr.right, target, seen);
    }

}
